Mathlib/Probability/Distributions/Fernique.lean

```diff
@@ -0,0 +1,602 @@
+  _ ≤ logRatio c * (a ^ 2 * (1 + √2) ^ 2 * 2 ^ (n + 2)) * a⁻¹ ^ 2 := by
+    gcongr
+    · exact (logRatio_pos hc_gt hc_lt).le
+    · exact sq_normThreshold_add_one_le n
+  _ = 2⁻¹ * Real.log (c.toReal / (1 - c).toReal) * 2 ^ n := by
+    unfold logRatio
+    field_simp
+    ring
+
+open Metric in
+lemma lintegral_closedBall_diff_exp_logRatio_mul_sq_le [IsProbabilityMeasure μ]
+    (h_rot : (μ.prod μ).map (ContinuousLinearMap.rotation (-(π / 4))) = μ.prod μ)
+    (ha_gt : 2⁻¹ < μ {x | ‖x‖ ≤ a}) (ha_lt : μ {x | ‖x‖ ≤ a} < 1) (n : ℕ) :
+    ∫⁻ x in (closedBall 0 (normThreshold a (n + 1)) \ closedBall 0 (normThreshold a n)),
+        .ofReal (rexp (logRatio (μ {x | ‖x‖ ≤ a}) * a⁻¹ ^ 2 * ‖x‖ ^ 2)) ∂μ
+      ≤ μ {x | ‖x‖ ≤ a} * .ofReal (rexp
+          (- 2⁻¹ * Real.log (μ {x | ‖x‖ ≤ a} / (1 - μ {x | ‖x‖ ≤ a})).toReal * 2 ^ n)) :=
+  let t := normThreshold a
+  let c := μ {x | ‖x‖ ≤ a}
+  let C := logRatio c * a⁻¹ ^ 2
+  calc ∫⁻ x in (closedBall 0 (t (n + 1)) \ closedBall 0 (t n)), .ofReal (rexp (C * ‖x‖ ^ 2)) ∂μ
+  _ ≤ ∫⁻ x in (closedBall 0 (t (n + 1)) \ closedBall 0 (t n)),
+      .ofReal (rexp (C * t (n + 1) ^ 2)) ∂μ := by
+    refine setLIntegral_mono (by fun_prop) fun x hx ↦ ?_
+    gcongr
+    · exact mul_nonneg (logRatio_pos ha_gt ha_lt).le (by positivity)
+    · simp only [Set.mem_diff, mem_closedBall, dist_zero_right, not_le] at hx
+      exact hx.1
+  _ = .ofReal (rexp (C * t (n + 1) ^ 2)) * μ (closedBall 0 (t (n + 1)) \ closedBall 0 (t n)) := by
+    simp only [lintegral_const, MeasurableSet.univ, Measure.restrict_apply, Set.univ_inter, C, t]
+  _ ≤ .ofReal (rexp (C * t (n + 1) ^ 2)) * μ {x | t n < ‖x‖} := by
+    gcongr
+    intro x
+    simp
+  _ ≤ .ofReal (rexp (C * t (n + 1) ^ 2))
+      * c * .ofReal (rexp (- Real.log (c / (1 - c)).toReal * 2 ^ n)) := by
+    conv_rhs => rw [mul_assoc]
+    gcongr
+    exact measure_gt_normThreshold_le_exp h_rot ha_gt ha_lt n
+  _ ≤ .ofReal (rexp (2⁻¹ * Real.log (c.toReal / (1 - c).toReal) * 2 ^ n))
+      * c * .ofReal (rexp (- Real.log (c / (1 - c)).toReal * 2 ^ n)) := by
+    gcongr ENNReal.ofReal (rexp ?_) * _ * _
+    convert logRatio_mul_normThreshold_add_one_le ha_gt ha_lt n (a := a) using 1
+    ring
+  _ = c * .ofReal (rexp (- 2⁻¹ * Real.log (c / (1 - c)).toReal * 2 ^ n)) := by
+    rw [mul_comm _ c, mul_assoc, ← ENNReal.ofReal_mul (by positivity), ← Real.exp_add]
+    congr
+    norm_cast
+    simp only [Nat.cast_pow, Nat.cast_ofNat, ENNReal.toReal_div, neg_mul]
+    field_simp
+    ring
+
+open Metric in
+lemma lintegral_exp_mul_sq_norm_le_mul [IsProbabilityMeasure μ]
+    (h_rot : (μ.prod μ).map (ContinuousLinearMap.rotation (-(π / 4))) = μ.prod μ)
+    (ha_pos : 0 < a)
+    {c' : ℝ≥0∞} (hc' : c' ≤ μ {x | ‖x‖ ≤ a}) (hc'_gt : 2⁻¹ < c') :
+    ∫⁻ x, .ofReal (rexp (logRatio c' * a⁻¹ ^ 2 * ‖x‖ ^ 2)) ∂μ
+      ≤ μ {x | ‖x‖ ≤ a} *
+       (.ofReal (rexp (logRatio c'))
+        + ∑' n, .ofReal (rexp (- 2⁻¹ * Real.log (c' / (1 - c')).toReal * 2 ^ n))) := by
+  let t := normThreshold a
```

Comment:
When a proof is that long, it helps to have a few comments explaining where the proof is currently, to follow the argument more easily.